lib/text/mp4_ttml_parser.js

```diff
@@ -80,67 +82,117 @@ shaka.text.Mp4TtmlParser = class {
+              box.reader, box.flags);
+          defaultSampleSize = parsedTFHDBox.defaultSampleSize;
+        })
+        .fullBox('trun', (box) => {
+          goog.asserts.assert(
+              box.version != null,
+              'A TRUN box should have a valid version value');
+          goog.asserts.assert(
+              box.flags != null,
+              'A TRUN box should have a valid flags value');
+
+          const parsedTRUNBox = shaka.util.Mp4BoxParsers.parseTRUN(
+              box.reader, box.version, box.flags);
+
+          for (const sample of parsedTRUNBox.sampleData) {
+            const sampleSize =
+                sample.sampleSize || defaultSampleSize || 0;
+            sampleSizes.push(sampleSize);
+          }
+        })
         .fullBox('subs', (box) => {
-          subSizes = [];
           const reader = box.reader;
           const entryCount = reader.readUint32();
+          let currentSampleNum = -1;
           for (let i = 0; i < entryCount; i++) {
-            reader.readUint32(); // sample_delta
+            const sampleDelta = reader.readUint32();
+            currentSampleNum += sampleDelta;
             const subsampleCount = reader.readUint16();
+            const subsampleSizes = [];
             for (let j = 0; j < subsampleCount; j++) {
               if (box.version == 1) {
-                subSizes.push(reader.readUint32());
+                subsampleSizes.push(reader.readUint32());
               } else {
-                subSizes.push(reader.readUint16());
+                subsampleSizes.push(reader.readUint16());
               }
               reader.readUint8(); // priority
               reader.readUint8(); // discardable
               reader.readUint32(); // codec_specific_parameters
             }
+            subSizesPerSample[currentSampleNum] = subsampleSizes;
           }
         })
         .box('mdat', Mp4Parser.allData((data) => {
-          sawMDAT = true;
-          // Join this to any previous payload, in case the mp4 has multiple
-          // mdats.
-          if (subSizes.length) {
-            const contentData =
-                shaka.util.BufferUtils.toUint8(data, 0, subSizes[0]);
-            const images = [];
-            let offset = subSizes[0];
-            for (let i = 1; i < subSizes.length; i++) {
-              const imageData =
-                  shaka.util.BufferUtils.toUint8(data, offset, subSizes[i]);
-              const raw =
-                  shaka.util.Uint8ArrayUtils.toStandardBase64(imageData);
-              images.push('data:image/png;base64,' + raw);
-              offset += subSizes[i];
-            }
-            payload = payload.concat(
-                this.parser_.parseMedia(contentData, time, uri, images));
-          } else {
-            payload = payload.concat(
-                this.parser_.parseMedia(data, time, uri, /* images= */ []));
-          }
+          // We first only collect all mdat, in case the mp4 has
```

Comment:
This comment is confusingly worded. I assume you meant something like this?

> We collect all of the mdats first, before parsing any of them. This is necessary in case the mp4 has multiple mdats.